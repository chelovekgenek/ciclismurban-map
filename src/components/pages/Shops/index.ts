export { ShopsContainer as Shops } from "./Shops.container"
