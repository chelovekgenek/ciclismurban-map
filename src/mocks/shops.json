[{ "point": { "lat": 47.0265823, "lng": 28.8180033 } }]
