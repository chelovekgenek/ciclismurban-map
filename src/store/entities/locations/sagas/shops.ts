import { takeLatest, put, call } from "redux-saga/effects"
import { AxiosResponse } from "axios"
import { extend } from "lodash-es"
import { ShopModel } from "@ciclismurban/models"

import { history } from "store/history"

import { getShops, deleteShop, updateShop, uploadFile, createShop } from "../api"
import {
  ShopsGetActions,
  ShopsGetTypes,
  ShopsCreateTypes,
  ShopsUpdateTypes,
  ShopsDeleteTypes,
  ShopsCreateActions,
  ShopsUpdateActions,
  ShopsDeleteActions,
  SelectedActions,
} from "../actions"

function* handleGet() {
  try {
    const { data }: AxiosResponse<ShopModel[]> = yield call(getShops)
    yield put(ShopsGetActions.success(data))
  } catch (e) {
    yield put(ShopsGetActions.failure(e))
  }
}

type THandleCreateAction = ReturnType<typeof ShopsCreateActions.request>
function* handleCreate({ payload: { image, ...payload } }: THandleCreateAction) {
  try {
    const { data: link }: AxiosResponse<string> = yield call(uploadFile, image!)
    const { data }: AxiosResponse<ShopModel> = yield call(createShop, { ...payload, image: link })
    yield put(ShopsCreateActions.success(data))
    yield call(history.replace, { pathname: "/shops" })
  } catch (e) {
    yield put(ShopsCreateActions.failure(e))
  }
}

type THandleUpdateAction = ReturnType<typeof ShopsUpdateActions.request>
function* handleUpdate({ payload: { uuid, payload } }: THandleUpdateAction) {
  try {
    const reqPayload = extend(payload, {
      image: payload.image ? (yield call(uploadFile, payload.image)).data : undefined,
    })
    const { data }: AxiosResponse<ShopModel> = yield call(updateShop, uuid, reqPayload)
    yield put(ShopsUpdateActions.success(data))
    yield put(SelectedActions.set(data))
  } catch (e) {
    yield put(ShopsUpdateActions.failure(e))
  }
}

type THandleDeleteAction = ReturnType<typeof ShopsDeleteActions.request>
function* handleDelete({ payload }: THandleDeleteAction) {
  try {
    const { data }: AxiosResponse<string> = yield call(deleteShop, payload)
    yield put(ShopsDeleteActions.success(data))
  } catch (e) {
    yield put(ShopsDeleteActions.failure(e))
  }
}

export function* watcher() {
  yield takeLatest(ShopsGetTypes.REQUEST, handleGet)
  yield takeLatest(ShopsCreateTypes.REQUEST, handleCreate)
  yield takeLatest(ShopsUpdateTypes.REQUEST, handleUpdate)
  yield takeLatest(ShopsDeleteTypes.REQUEST, handleDelete)
}
