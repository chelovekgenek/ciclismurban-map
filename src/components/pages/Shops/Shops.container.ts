import { connect } from "react-redux"

import { Actions as ShopsActions, Selectors as ShopsSelectors } from "store/entities/locations/shops"
import { TAppState } from "store/entities"

import { Shops } from "./Shops"

export interface IStateProps {
  shops: ReturnType<typeof ShopsSelectors.getLocations>
  fetching: ReturnType<typeof ShopsSelectors.getFetching>
}
export interface IDispatchProps {
  getShops: typeof ShopsActions.Get.request
  deleteShop: typeof ShopsActions.Delete.request
}

export const ShopsContainer = connect<IStateProps, IDispatchProps, {}, TAppState>(
  state => ({
    shops: ShopsSelectors.getLocations(state),
    fetching: ShopsSelectors.getFetching(state),
  }),
  {
    getShops: ShopsActions.Get.request,
    deleteShop: ShopsActions.Delete.request,
  },
)(Shops)
