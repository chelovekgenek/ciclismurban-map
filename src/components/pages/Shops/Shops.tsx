import React from "react"

import { Locations } from "components/generic/page"

import { IStateProps, IDispatchProps } from "./Shops.container"

interface IProps extends IStateProps, IDispatchProps {}

export const Shops: React.FC<IProps> = ({ shops, fetching, getShops, deleteShop }) => (
  <Locations
    title="Магазины"
    fetching={fetching}
    locations={shops}
    links={{ create: "/shop", getInfo: uuid => `/shop/${uuid}` }}
    getLocations={getShops}
    deleteLocation={deleteShop}
  />
)
