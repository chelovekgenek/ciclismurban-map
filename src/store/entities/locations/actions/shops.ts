import { IState as IShopsState } from "../reducers/shops"
import { action, payload } from "ts-action"

export enum ShopsTypes {
  GET__REQUEST = "LOCATIONS__SHOPS__GET__REQUEST",
  GET__SUCCESS = "LOCATIONS__SHOPS__GET__SUCCESS",
  GET__FAILURE = "LOCATIONS__SHOPS__GET__FAILURE",
}

type TRequiredShopsState = Required<IShopsState>
export const ShopsActions = {
  requestGet: action(ShopsTypes.GET__REQUEST),
  successGet: action(ShopsTypes.GET__SUCCESS, payload<TRequiredShopsState["data"]>()),
  failureGet: action(ShopsTypes.GET__FAILURE, payload<TRequiredShopsState["error"]>()),
}
