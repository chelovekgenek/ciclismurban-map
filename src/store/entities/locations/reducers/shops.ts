import { reducer, on, union } from "ts-action"
import { ShopModel } from "@ciclismurban/models"

import { ShopsGetActions, ShopsCreateActions, ShopsUpdateActions, ShopsDeleteActions } from "../actions"

export interface IState {
  data: ShopModel[]
  fetching: boolean
  error?: Error
}

const initialState: IState = {
  data: [],
  fetching: false,
}

export default reducer(
  initialState,
  on(
    ...union(
      ShopsGetActions.request,
      ShopsCreateActions.request,
      ShopsUpdateActions.request,
      ShopsDeleteActions.request,
    ),
    state => ({
      ...state,
      fetching: true,
    }),
  ),
  on(
    ...union(
      ShopsGetActions.failure,
      ShopsCreateActions.failure,
      ShopsUpdateActions.failure,
      ShopsDeleteActions.failure,
    ),
    (state, { payload }) => ({
      ...state,
      fetching: false,
      error: payload,
    }),
  ),
  on(ShopsGetActions.success, (state, { payload }) => ({ ...state, fetching: false, data: payload })),
  on(ShopsCreateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.concat(payload),
  })),
  on(ShopsUpdateActions.success, (state, { payload }) => ({
    ...state,
    fetching: false,
    data: state.data.map(item => (item.uuid === payload.uuid ? payload : item)),
  })),
  on(ShopsDeleteActions.success, (state, { payload }) => ({
    ...state,
    data: state.data.filter(item => item.uuid !== payload),
    fetching: false,
  })),
)
