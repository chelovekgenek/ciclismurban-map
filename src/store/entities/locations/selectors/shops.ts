import { TAppState } from "store/entities/reducers"

export const getShopsFetching = (state: TAppState) => state.locations.shops.fetching
export const getShopsLocations = (state: TAppState) => state.locations.shops.data
