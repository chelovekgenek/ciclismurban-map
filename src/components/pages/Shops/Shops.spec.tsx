import React from "react"
import { shallow } from "enzyme"

import shops from "mocks/shops.json"

import { Shops } from "./Shops"

describe("pages/Shops", () => {
  const render = (props: Partial<React.ComponentProps<typeof Shops>> = {}) =>
    shallow(
      <Shops
        shops={shops as any}
        fetching={false}
        getShops={jest.fn() as any}
        deleteShop={jest.fn() as any}
        {...props}
      />,
    )
  it("should match snapshot", () => {
    expect(render()).toMatchSnapshot()
  })
})
