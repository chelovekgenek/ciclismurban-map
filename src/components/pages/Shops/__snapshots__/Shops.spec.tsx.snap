// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Shops should match snapshot 1`] = `
<withRouter(Locations)
  deleteLocation={[MockFunction]}
  fetching={false}
  getLocations={[MockFunction]}
  links={
    Object {
      "create": "/shop",
      "getInfo": [Function],
    }
  }
  locations={
    Array [
      Object {
        "createdAt": "2019-07-31T23:22:49.173Z",
        "description": "Et harum quidem rerum facilis est et expedita distinctio. Nam libero tempore, cum soluta nobis est eligendi optio cumque nihil impedit quo minus id quod maxime placeat facere possimus, omnis voluptas assumenda est, omnis dolor repellendus. Temporibus autem quibusdam et aut officiis debitis aut rerum necessitatibus saepe eveniet ut et voluptates repudiandae sint et molestiae non recusandae. Itaque earum rerum hic tenetur a sapiente delectus, ut aut reiciendis voluptatibus maiores alias consequatur aut perferendis doloribus asperiores repellat.",
        "image": "https://cdn.shopify.com/s/files/1/2340/0283/files/TreadlyShopHeader1_1400x.progressive.jpg?v=1508754615",
        "point": Object {
          "lat": 47.0265823,
          "lng": 28.8180033,
        },
        "title": "ShopA",
        "updatedAt": "2019-07-31T23:22:49.173Z",
        "uuid": "fbf27a68-4af3-4efb-8f48-e64dd1cf0807",
      },
    ]
  }
  title="Магазины"
/>
`;
